apps/rush-lib/src/RushX.ts

```diff
@@ -0,0 +1,134 @@
+import * as path from 'path';
+
+import {
+  PackageJsonLookup,
+  IPackageJson,
+  Text,
+  IPackageJsonScriptTable
+ } from '@microsoft/node-core-library';
+import { Utilities } from '../lib/utilities/Utilities';
+
+export class RushX {
+
+  public static launchRushX(launcherVersion: string, isManaged: boolean): void {
+    // NodeJS can sometimes accidentally terminate with a zero exit code  (e.g. for an uncaught
+    // promise exception), so we start with the assumption that the exit code is 1
+    // and set it to 0 only on success.
+    process.exitCode = 1;
+
+    try {
+      // Find the governing package.json for this folder:
+      const packageJsonLookup: PackageJsonLookup = new PackageJsonLookup();
+
+      const packageJsonFilePath: string | undefined = packageJsonLookup.tryGetPackageJsonFilePathFor(process.cwd());
+      if (!packageJsonFilePath) {
+        console.log(colors.red('This command should be used inside a project folder.'));
+        console.log(`Unable to find a package.json file in the current working directory or any of its parents.`);
+        return;
+      }
+
+      const packageJson: IPackageJson = packageJsonLookup.loadPackageJson(packageJsonFilePath);
+
+      // 0 = node.exe
+      // 1 = rushx
+      const args: string[] = process.argv.slice(2);
+
+      if (args.length === 0 || args[0][0] === '-') {
+        RushX._showUsage(packageJson);
+        return;
+      }
+
+      const command: string = args[0];
+
+      const scripts: IPackageJsonScriptTable = packageJson.scripts || { };
+
+      if (!Object.hasOwnProperty.call(scripts, command)) {
+        console.log(colors.red(`Error: The command "${command}" is not defined in the`
+          + ` package.json file for this project.`));
+
+        const availableCommands: string[] = Object.keys(scripts);
+        if (availableCommands.length > 0) {
+          console.log(os.EOL + 'Available commands for this project are: '
+            + availableCommands.map(x => `"${x}"`).join(', '));
+        }
+
+        console.log(`Use ${colors.yellow('"rushx --help"')} for more information.`);
+        return;
+      }
+
+      const scriptBody: string = scripts[command];
+      console.log('Executing: ' + JSON.stringify(scriptBody) + os.EOL);
+
+      const packageFolder: string = path.dirname(packageJsonFilePath);
+
+      const result: child_process.SpawnSyncReturns<Buffer> = Utilities.executeLifecycleCommand(
+        scriptBody,
+        packageFolder,
+        packageFolder
+      );
+
+      // Return the exit code of the child process
+      process.exitCode = result.status;
+    } catch (error) {
+      console.log(colors.red('Error: ' + error.message));
+    }
+  }
+
+  private static _showUsage(packageJson: IPackageJson): void {
+    console.log('usage: rushx [-h]');
+    console.log('       rushx <command> ...' + os.EOL);
+
+    console.log('Optional arguments:');
+    console.log('  -h, --help            Show this help message and exit.' + os.EOL);
```

Comment:
Can this use ts-command-line?

Reply:
It can't because we don't know what the command line is.  Rush Stack defines a project-commands.json that allows you to fully document your command line; in the future, for projects using that approach, we will use ts-command-line

---
In reply to: [193581089](https://github.com/Microsoft/web-build-tools/pull/678#discussion_r193581089) [](ancestors = 193581089)